src/control/lib/hardware/topology.go

```diff
@@ -34,20 +41,130 @@ func (t *Topology) AllDevices() map[string]*Device {
 	return devsByName
 }
 
+func PrintTopology(t *Topology, output io.Writer) error {
+	if t == nil {
+		fmt.Fprintf(output, "No topology information available\n")
+	}
+
+	for _, numaNode := range t.NUMANodes {
+		coreSet := &system.RankSet{}
+		for _, core := range numaNode.Cores {
+			coreSet.Add(system.Rank(core.ID))
+		}
+
+		fmt.Fprintf(output, "NUMA node %d (%s)\n", numaNode.ID, numaNode.Set)
+		fmt.Fprintf(output, "  CPU cores: %s\n", coreSet)
+		fmt.Fprintf(output, "  PCI devices: \n")
+		for addr, devs := range numaNode.PCIDevices {
+			fmt.Fprintf(output, "    %s\n", &addr)
+			for _, dev := range devs {
+				fmt.Fprintf(output, "      %s\n", dev)
+			}
+		}
+		fmt.Fprintf(output, "  PCI buses: \n")
+		for _, bus := range numaNode.PCIBuses {
+			fmt.Fprintf(output, "    %s\n", bus)
+		}
+	}
+
+	return nil
+}
+
 type (
+	CPUCore struct {
+		ID       uint      `json:"id"`
+		NUMANode *NUMANode `json:"-"`
+	}
+
+	// PCIDevice represents an individual hardware device.
+	PCIDevice struct {
+		Name      string            `json:"name"`
+		Type      DeviceType        `json:"type"`
+		NUMANode  *NUMANode         `json:"-"`
+		PCIAddr   common.PCIAddress `json:"pci_address"`
+		LinkSpeed float64           `json:"link_speed"`
+	}
+
+	PCIBus struct {
```

Comment:
I wonder if there's some value to having a PCIDevices member in the PCIBus, containing the associated devices. I am doing data structures like that in the fabric stuff, making it possible to look up items in different ways quickly. I'm using pointers so it's not taking too much extra memory.

Reply:
I'll see if I can make that work.